core/node/logs_bloom_backfill/src/lib.rs

```diff
@@ -0,0 +1,113 @@
+use std::time::Duration;
+
+use anyhow::Context;
+use tokio::sync::watch;
+use zksync_dal::{Connection, ConnectionPool, Core, CoreDal};
+use zksync_types::{block::build_bloom, BloomInput, L2BlockNumber};
+
+#[derive(Debug)]
+pub struct LogsBloomBackfill {
+    connection_pool: ConnectionPool<Core>,
+}
+
+impl LogsBloomBackfill {
+    pub fn new(connection_pool: ConnectionPool<Core>) -> Self {
+        Self { connection_pool }
+    }
+
+    async fn wait_for_l2_block_with_bloom(
+        connection: &mut Connection<'_, Core>,
+        stop_receiver: &mut watch::Receiver<bool>,
+    ) -> anyhow::Result<bool> {
+        const INTERVAL: Duration = Duration::from_secs(1);
+        tracing::debug!("waiting for at least one L2 block in DB with bloom");
+
+        loop {
+            if *stop_receiver.borrow() {
+                return Ok(false);
+            }
+
+            let bloom = connection.blocks_dal().get_last_l2_block_bloom().await?;
+            if bloom.is_some() {
+                return Ok(true);
+            }
+
+            // We don't check the result: if a stop signal is received, we'll return at the start
+            // of the next iteration.
+            tokio::time::timeout(INTERVAL, stop_receiver.changed())
+                .await
+                .ok();
+        }
+    }
+
+    pub async fn run(self, mut stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        let mut connection = self
+            .connection_pool
+            .connection_tagged("logs_bloom_backfill")
+            .await?;
+
+        if !Self::wait_for_l2_block_with_bloom(&mut connection, &mut stop_receiver).await? {
+            return Ok(()); // Stop signal received
+        }
+
+        let max_block_without_bloom = connection
+            .blocks_dal()
+            .get_max_l2_block_without_bloom()
+            .await?;
+        let Some(max_block_without_bloom) = max_block_without_bloom else {
+            tracing::info!("all blooms are already there, exiting migration");
+            return Ok(());
+        };
+        let first_l2_block = connection
+            .blocks_dal()
+            .get_earliest_l2_block_number()
+            .await?
+            .context(
+                "logs_bloom_backfill: missing l2 block in DB after waiting for at least one",
+            )?;
+
+        tracing::info!("starting blooms backfill");
```

Comment:
Nit: IMO, it would make sense to log the range of filled blocks (maybe even have tracing spans for the backfill procedure and each `WINDOW`).

Reply:
I've added more logs, IMO spans are an overkill in this case